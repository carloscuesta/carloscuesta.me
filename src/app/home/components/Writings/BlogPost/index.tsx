import Image from 'next/image'
import Link from 'next/link'

import { type PostPreview } from 'src/utils/api/blog/mutators'

type Props = {
  post: PostPreview
  isImageLoadedWithPriority: boolean
}

const BlogPost = (props: Props) => (
  <div className="w-full flex-none snap-start px-2 py-2 sm:w-1/2 md:w-1/3">
    <Link
      href={`/blog/${props.post.slug}`}
      title={props.post.title}
      className="h-full rounded-lg outline-none ring-offset-4 focus-visible:ring-2 sm:flex"
    >
      <article>
        <Image
          alt=""
          blurDataURL={props.post.images.preview.lqpi}
          className="mb-3 w-full rounded-lg"
          height={150}
          placeholder="blur"
          priority={props.isImageLoadedWithPriority}
          src={props.post.images.preview.src}
          width={300}
        />

        <div className="grid gap-2">
          <header className="font-semibold">{props.post.title}</header>

          <p className="text-sm opacity-70">{props.post.excerpt}</p>
        </div>
      </article>
    </Link>
  </div>
)

export default BlogPost
