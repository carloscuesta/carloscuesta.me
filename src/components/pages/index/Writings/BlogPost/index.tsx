import Link from 'next/link'

import { type PostPreview } from 'src/utils/api/blog/mutators'

type Props = {
  post: PostPreview
}

const BlogPost = (props: Props) => (
  <div className="w-full sm:w-1/2 md:w-1/3 flex-none snap-start px-2 py-2">
    <Link
      href={`/blog/${props.post.slug}`}
      title={props.post.title}
      className='sm:flex rounded-lg h-full outline-none focus-visible:ring-2 ring-offset-4'
    >
      <article>
        <img
          className='w-full lazyload rounded-lg mb-3'
          data-src={props.post.images.preview.src}
          src={props.post.images.preview.lqpi}
        />

        <div className='grid gap-2'>
          <header className='font-semibold'>
            {props.post.title}
          </header>

          <p className='opacity-70 text-sm'>
            {props.post.excerpt}
          </p>
        </div>
      </article>
    </Link>
  </div>
  
)

export default BlogPost
