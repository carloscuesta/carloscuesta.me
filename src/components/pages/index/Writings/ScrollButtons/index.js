// @flow
import type { Element } from 'react'

import styles from './styles.module.css'

type Props = { scrollTo: Function, scrollPosition: number, scrollPositionMaxWidth: number }

const SCROLL_VISIBILITY_THRESHOLD: number = 1

const ScrollButtons = (props: Props): Element<'div'> => (
  <div className='row'>
    <div className={`col-xs-12 ${styles.scrollButtons}`}>
      <button
        className={styles.button}
        disabled={props.scrollPosition <= SCROLL_VISIBILITY_THRESHOLD}
        onClick={() => props.scrollTo('previous')}
      >
        <svg
          fill='currentColor'
          height='24'
          width='24'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path d='M15.422 7.406L10.828 12l4.594 4.594L14.016 18l-6-6 6-6z' />
        </svg>
        <span>Back</span>
      </button>

      <button
        className={styles.button}
        disabled={props.scrollPosition >= (props.scrollPositionMaxWidth - SCROLL_VISIBILITY_THRESHOLD)}
        onClick={() => props.scrollTo('next')}
      >
        <span>Next</span>
        <svg
          fill='currentColor'
          height='24'
          width='24'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path d='M9.984 6l6 6-6 6-1.406-1.406L13.172 12 8.578 7.406z' />
        </svg>
      </button>
    </div>
  </div>
)

export default ScrollButtons
