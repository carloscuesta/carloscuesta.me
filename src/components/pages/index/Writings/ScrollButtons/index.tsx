type Props = {
  scrollTo: (action: 'next' | 'previous') => void,
  scrollPosition: number,
  scrollPositionMaxWidth: number
}

const SCROLL_VISIBILITY_THRESHOLD = 1

const ScrollButtons = (props: Props) => (
  <div className='flex justify-between mt-2'>
    <button
      className='flex touch-manipulation cursor-pointer items-center disabled:invisible text-sm mt-2 opacity-70 hover:opacity-100 transition-opacity'
      disabled={props.scrollPosition <= SCROLL_VISIBILITY_THRESHOLD}
      onClick={() => props.scrollTo('previous')}
    >
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" className="w-5 h-5">
        <path fillRule="evenodd" d="M12.79 5.23a.75.75 0 01-.02 1.06L8.832 10l3.938 3.71a.75.75 0 11-1.04 1.08l-4.5-4.25a.75.75 0 010-1.08l4.5-4.25a.75.75 0 011.06.02z" clipRule="evenodd" />
      </svg>
      <span>Back</span>
    </button>

    <button
      className='flex touch-manipulation cursor-pointer items-center disabled:invisible text-sm opacity-70 hover:opacity-100 transition-opacity'
      disabled={props.scrollPosition >= (props.scrollPositionMaxWidth - SCROLL_VISIBILITY_THRESHOLD)}
      onClick={() => props.scrollTo('next')}
    >
      <span>Next</span>
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" className="w-5 h-5">
        <path fillRule="evenodd" d="M7.21 14.77a.75.75 0 01.02-1.06L11.168 10 7.23 6.29a.75.75 0 111.04-1.08l4.5 4.25a.75.75 0 010 1.08l-4.5 4.25a.75.75 0 01-1.06-.02z" clipRule="evenodd" />
      </svg>
    </button>
  </div>
)

export default ScrollButtons
