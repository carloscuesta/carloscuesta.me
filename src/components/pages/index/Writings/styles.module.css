.scrollablePosts {
  /* @FIXME: I need to set !important to override this property on Firefox
    this overrides the row value.
  */
  flex-wrap: nowrap!important;
  overflow-x: scroll;
  scroll-snap-type: x mandatory;
}

.scrollablePosts:global(::-webkit-scrollbar) {
  -webkit-appearance: none;
  height: .5em;
}

.scrollablePosts:global(::-webkit-scrollbar-thumb) {
  border-radius: 4px;
  background-color: var(--textSecondary);
}

.scrollablePosts:global(::-webkit-scrollbar-track) {
  border-radius: 4px;
  box-shadow: 0 1px 4px var(--shadowColor);
  margin: 0 1.15em;
}

@media (prefers-color-scheme: dark) {
  .scrollablePosts:global(::-webkit-scrollbar-track) {
    box-shadow: none;
  }
}
