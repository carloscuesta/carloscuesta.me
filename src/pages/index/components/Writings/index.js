// @flow
import React from 'react'

type Props = {}

const Writings = (props: Props) => (
  <>
    {props.posts.map((post) => (
      <h1 key={post.url}>{post.title}</h1>
    ))}
  </>
)

export default Writings
