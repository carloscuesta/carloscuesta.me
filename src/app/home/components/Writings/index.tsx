'use client'

import { useRef, useState } from 'react'
import debounce from 'lodash.debounce'

import { type PostPreview } from 'src/utils/api/blog/mutators'
import Wrapper from 'src/components/Wrapper'
import SectionTitle from 'src/components/SectionTitle'
import ScrollButtons from './ScrollButtons'
import BlogPost from './BlogPost'

type Props = {
  posts: Array<PostPreview>
}

const Writings = (props: Props) => {
  const scrollablePostsRef = useRef<HTMLDivElement>(null)
  const [scrollPosition, setScrollPosition] = useState(0)
  const SCROLL_DEBOUNCE_MS_TIME = 20

  const onScroll = debounce(
    (event) => setScrollPosition(event.target.scrollLeft),
    SCROLL_DEBOUNCE_MS_TIME
  )

  const scrollTo = (action: 'next' | 'previous') => {
    if (!scrollablePostsRef.current) return

    const postWidth = (
      scrollablePostsRef.current.childNodes[0] as HTMLDivElement
    ).offsetWidth
    const scrollPosition =
      action === 'next'
        ? scrollablePostsRef.current.scrollLeft + postWidth
        : scrollablePostsRef.current.scrollLeft - postWidth

    scrollablePostsRef.current.scrollTo({
      behavior: 'smooth',
      left: scrollPosition,
      top: 0,
    })
  }

  return (
    <section>
      <style>
        {`
          .scrollablePosts::-webkit-scrollbar {
            -webkit-appearance: none;
            display: none;
          }
        `}
      </style>
      <Wrapper>
        <SectionTitle
          subTitle="The latests posts of my blog."
          title="Writings"
          viewAllLink="/blog"
        />

        <div
          data-testid="scrollablePosts"
          className="scrollablePosts -mx-2 -my-2 flex snap-x snap-mandatory overflow-x-auto"
          onScroll={onScroll}
          ref={scrollablePostsRef}
        >
          {props.posts.map((post, index) => (
            <BlogPost
              isImageLoadedWithPriority={[0, 1, 2].includes(index)}
              key={post.slug}
              post={post}
            />
          ))}
        </div>

        <ScrollButtons
          scrollTo={scrollTo}
          scrollPosition={scrollPosition}
          scrollPositionMaxWidth={
            Number(scrollablePostsRef?.current?.scrollWidth) -
            Number(scrollablePostsRef?.current?.clientWidth)
          }
        />
      </Wrapper>
    </section>
  )
}

export default Writings
