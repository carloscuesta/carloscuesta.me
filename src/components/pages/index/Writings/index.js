// @flow
import React, { type Element } from 'react'

import { type PostPreview } from 'src/utils/api/blog/mutators'
import Wrapper from 'src/components/shared/Wrapper'
import BlogPost from 'src/components/shared/BlogPost'
import SectionTitle from 'src/components/shared/SectionTitle'
import ScrollButtons from './ScrollButtons'
import styles from './styles.module.css'

type Props = {
  posts: Array<PostPreview>
}

const Writings = (props: Props): Element<'section'> => {
  const scrollablePostsRef: Object = React.useRef({ current: {} })
  const [postWidth, setPostWidth] = React.useState(0)

  React.useEffect(() => {
    setPostWidth(scrollablePostsRef.current.childNodes[0].offsetWidth)
  }, [])

  const scrollTo = (action: 'next' | 'previous') => {
    setPostWidth(scrollablePostsRef.current.childNodes[0].offsetWidth)

    scrollablePostsRef.current.scrollTo({
      behavior: 'smooth',
      left: action === 'next'
        ? scrollablePostsRef.current.scrollLeft + postWidth
        : scrollablePostsRef.current.scrollLeft - postWidth,
      top: 0
    })
  }

  return (
    <section>
      <Wrapper>
        <SectionTitle
          isCentered
          subTitle='The latests posts of my blog.'
          title='Writings'
        />

        <div ref={scrollablePostsRef} className={`row ${styles.scrollablePosts}`}>
          {props.posts.map((post) => (
            <BlogPost key={post.slug} post={post} />
          ))}
        </div>

        <ScrollButtons scrollTo={scrollTo} />
      </Wrapper>
    </section>
  )
}

export default Writings
