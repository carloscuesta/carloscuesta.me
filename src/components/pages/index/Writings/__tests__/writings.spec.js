import renderer from 'react-test-renderer'

import Writings from '../index'
import * as stubs from './stubs'

describe('Writings', () => {
  it('should match Writings component', () => {
    const wrapper = renderer.create(<Writings {...stubs.props} />)

    expect(wrapper).toMatchSnapshot()
  })

  describe('ScrollButtons', () => {
    describe('on scroll', () => {
      describe('when action is next', () => {
        it('should call scrollTo and move to the next scroll position', () => {
          const scrollablePostsRef = stubs.scrollablePostsRefMock()
          const wrapper = renderer.create(
            <Writings {...stubs.props} />,
            { createNodeMock: () => scrollablePostsRef }
          )

          renderer.act(() => {
            wrapper.root.findAllByType('button')[1].props.onClick()
          })

          expect(scrollablePostsRef.scrollTo).toHaveBeenCalledWith({
            behavior: 'smooth',
            top: 0,
            left: scrollablePostsRef.childNodes[0].offsetWidth
          })
        })
      })

      describe('when action is previous', () => {
        it('should call scrollTo and move to the previous scroll position', () => {
          const scrollablePostsRef = stubs.scrollablePostsRefMock()
          const wrapper = renderer.create(
            <Writings {...stubs.props} />,
            { createNodeMock: () => scrollablePostsRef }
          )

          renderer.act(() => {
            wrapper.root.findAllByType('button')[1].props.onClick()
            wrapper.root.findAllByType('button')[0].props.onClick()
          })

          expect(scrollablePostsRef.scrollTo).toHaveBeenCalledWith({
            behavior: 'smooth',
            top: 0,
            left: -scrollablePostsRef.childNodes[0].offsetWidth
          })
        })
      })
    })

    describe('when user has not scrolled', () => {
      it('should disable the Back button', () => {
        const wrapper = renderer.create(<Writings {...stubs.props} />)

        expect(wrapper.root.findAllByType('button')[0].props.disabled).toBe(true)
      })
    })

    describe('when user has scrolled', () => {
      beforeEach(() => {
        jest.clearAllTimers()
      })

      describe('when the scrollPosition is equal to 0', () => {
        it('should disable the Back button', () => {
          const scrollablePostsRef = stubs.scrollablePostsRefMock()
          const wrapper = renderer.create(
            <Writings {...stubs.props} />,
            { createNodeMock: () => scrollablePostsRef }
          )

          expect(wrapper.root.findAllByType('button')[0].props.disabled).toBe(true)
        })
      })

      describe('when the scrollPosition is bigger than 0', () => {
        it('should enable the Back button', () => {
          const scrollablePostsRef = stubs.scrollablePostsRefMock()
          const wrapper = renderer.create(
            <Writings {...stubs.props} />,
            { createNodeMock: () => scrollablePostsRef }
          )

          renderer.act(() => {
            jest.useFakeTimers()
            wrapper.root.findByProps({ className: 'row scrollablePosts' }).props.onScroll(
              stubs.scrollEventMock(350)
            )
            jest.runAllTimers()
          })

          expect(wrapper.root.findAllByType('button')[0].props.disabled).toBe(false)
        })
      })

      describe('when the scrollPosition is equal to scrollPositionMaxWidth', () => {
        it('should disable the Next button', () => {
          const scrollablePostsRef = {
            ...stubs.scrollablePostsRefMock(),
            scrollWidth: 1350,
            clientWidth: 1000
          }

          const wrapper = renderer.create(
            <Writings {...stubs.props} />,
            { createNodeMock: () => scrollablePostsRef }
          )

          renderer.act(() => {
            jest.useFakeTimers()
            wrapper.root.findByProps({ className: 'row scrollablePosts' }).props.onScroll(
              stubs.scrollEventMock(scrollablePostsRef.scrollWidth - scrollablePostsRef.clientWidth)
            )
            jest.runAllTimers()
          })

          expect(wrapper.root.findAllByType('button')[1].props.disabled).toBe(true)
        })
      })
    })
  })
})
