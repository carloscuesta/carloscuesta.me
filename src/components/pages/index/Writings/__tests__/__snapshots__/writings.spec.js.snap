// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Writings should match Writings component 1`] = `
<section>
  <div
    className="wrapper"
  >
    <div
      className="container containerWithLink"
    >
      <div>
        <h2
          className="sectionTitle "
        >
          Writings
        </h2>
        <h3
          className="sectionSubtitle"
        >
          The latests posts of my blog.
        </h3>
      </div>
      <a
        href="/blog"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        View all
      </a>
    </div>
    <div
      className="row scrollablePosts"
      onScroll={[Function]}
    >
      <div
        className="col-xs-12 col-sm-6 col-md-4 postContainer"
      >
        <article
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <img
            alt="Post Title"
            className="lazyload postImage"
            data-src="https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1454754838/kcuq34zbo1yieygmx0y6.png"
            src="https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/v1454754838/kcuq34zbo1yieygmx0y6.png"
          />
          <div
            className="postContent"
          >
            <header
              className="postTitle"
            >
              Post Title
            </header>
            <time
              className="postPublishedAt"
              dateTime="2019-12-15 10:00"
            >
              A year ago
            </time>
            <p
              className="postExcerpt"
            >
              This is the post excerpt that will be shown.
            </p>
          </div>
        </article>
      </div>
      <div
        className="col-xs-12 col-sm-6 col-md-4 postContainer"
      >
        <article
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <img
            alt="A post"
            className="lazyload postImage"
            data-src="https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1454754838/kcuq34zbo1yieygmx0y6.png"
            src="https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/v1454754838/kcuq34zbo1yieygmx0y6.png"
          />
          <div
            className="postContent"
          >
            <header
              className="postTitle"
            >
              A post
            </header>
            <time
              className="postPublishedAt"
              dateTime="2019-12-17 10:00"
            >
              A year ago
            </time>
            <p
              className="postExcerpt"
            >
              This is the post excerpt that will be shown.
            </p>
          </div>
        </article>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-xs-12 scrollButtons"
      >
        <button
          className="button"
          disabled={true}
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.422 7.406L10.828 12l4.594 4.594L14.016 18l-6-6 6-6z"
            />
          </svg>
          <span>
            Back
          </span>
        </button>
        <button
          className="button"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            Next
          </span>
          <svg
            fill="currentColor"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.984 6l6 6-6 6-1.406-1.406L13.172 12 8.578 7.406z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</section>
`;
